src/ILCompiler.Compiler/src/Compiler/Compilation.cs

```diff
@@ -200,9 +202,53 @@ private void AddCompilationRoots()
                     AddCompilationRootsForMainMethod(module);
 
                 AddCompilationRootsForRuntimeExports(module);
-           }
+            }
+
+            if (_options.MultiFile && StartupCodeMain == null)
+            {
+                foreach (var inputFile in _typeSystemContext.InputFilePaths)
+                {
+                    var module = _typeSystemContext.GetModuleFromPath(inputFile.Value);
+
+                    foreach (TypeDesc type in module.GetAllTypes())
```

Comment:
Nit: I would put this into a separate method (e.g. `AddCompilationRootsForMultifileLibrary`) and put the if check followed by a call to this method in the foreach loop above.